File: solutions/Sorting/0295-find-median-from-data-stream/find-median-from-data-stream.cpp
class MedianFinder {
public:
    priority_queue<int> maxpq;
    priority_queue<int,vector<int>,greater<int>> minpq;
    
    void rebalance(){
        // maintain n1>=n2
        while(maxpq.size()<minpq.size()){
            maxpq.push(minpq.top());
            minpq.pop();
        }
        if(maxpq.size()-minpq.size()>1){
            minpq.push(maxpq.top());
            maxpq.pop();
        }
    }
    
    MedianFinder() {
        
    }
    
    void addNum(int num) {
        if(maxpq.empty() || num<=maxpq.top()){
            maxpq.push(num);
        }else{
            minpq.push(num);
        }
        rebalance();
    }
    
    double findMedian() {
        if(maxpq.size()==minpq.size()){
            return (maxpq.top()+minpq.top())/(2.0);
        }else{
            return maxpq.top();
        }
    }
};

/**
 * Your MedianFinder object will be instantiated and called as such:
 * MedianFinder* obj = new MedianFinder();
 * obj->addNum(num);
 * double param_2 = obj->findMedian();
 */

File: solutions/295-find-median-from-data-stream/find-median-from-data-stream.java
class MedianFinder {

    private final Queue<Integer> maxHeap;
    private final Queue<Integer> minHeap;

    public MedianFinder() {
        maxHeap=new PriorityQueue<>(Comparator.reverseOrder());
        minHeap=new PriorityQueue<>();
    }
    
    void rebalance(){
        
        if(maxHeap.size()>minHeap.size()+1){
            minHeap.offer(maxHeap.poll());
        }else if(maxHeap.size()<minHeap.size()){
            maxHeap.offer(minHeap.poll());
        }
    }

    public void addNum(int num) {
        if(maxHeap.isEmpty() || maxHeap.peek()>=num){
            maxHeap.offer(num);
        }else{
            minHeap.offer(num);
        }
        rebalance();
    }
    
    public double findMedian() {
        if(maxHeap.size()==minHeap.size()){
            return ((1.0)*maxHeap.peek()+minHeap.peek())/2;
        }else{
            return 1.0*(maxHeap.peek());
            
        }
    }
}

/**
 * Your MedianFinder object will be instantiated and called as such:
 * MedianFinder obj = new MedianFinder();
 * obj.addNum(num);
 * double param_2 = obj.findMedian();
 */